lib/cudnn/README.md

```diff
@@ -0,0 +1,90 @@
+Julia functions. Here are some design choices I followed:
+
+**Naming:** We try to keep the same function, argument, and type names from the cuDNN
+library in the high level interface. The wrappers for descriptors drop the `_t` suffix,
+e.g. `cudnnPoolingDescriptor_t => cudnnPoolingDescriptor`.
+
+**Descriptors:** The cuDNN functions take data and operator descriptors. Most of these
+descriptors are relatively fast to create (~500 ns for a cudnnTensorDescriptor) so they may
+not be worth preallocating for the user but we provide keyword options anyway. We cache
+descriptors (~100 ns) so we can use them as hash keys for memoization, which also saves a
+bit of memory and speed.  All descriptor fields are `isbits` types with the exception of the
+`cudnnDropoutDescriptor` which points to a random number generator state and is used as a
+field of some other descriptors.
+
+**Operator descriptors:** Descriptors such as `cudnnPoolingDescriptor` specify the options
+for an operator such as stride and padding. For operators with descriptors we have one
+method that takes keyword arguments with reasonable defaults to construct the descriptor and
+another method that takes a pre-initialized descriptor as its last argument.  This way a
+casual user can call the first method without worrying about the descriptor format, only
+specifying non-default options, whereas a layer architect can keep a preset descriptor in
+the layer that gets passed to the function using the second method. We try to use generic
+Julia types for keyword arguments that specify default descriptor fields and convert these
+to the appropriate cudnn types during descriptor construction.
+
+**Output arrays:** The low level cuDNN functions take pre-allocated output arrays. The high
+level interface has one Julia function that allocates its own output array
+(e.g. `cudnnPoolingForward`) and another with an exclamation mark that takes a pre-allocated
+output array as its first argument (e.g. `cudnnPoolingForward!`).
+
+**Methods:** Each cuDNN forward function may have up to four methods depending on whether
+the descriptor and the output array are specified:
+
+    cudnnPoolingForward(x; kwargs...)
+    cudnnPoolingForward(x, d::cudnnPoolingDescriptor; kwargs...)
+    cudnnPoolingForward!(y, x; kwargs...)
+    cudnnPoolingForward!(y, x, d::cudnnPoolingDescriptor; kwargs...)
+
+The conventional order of arguments for these public methods is:
+
+    ([output], weights, inputs, [descriptor]; kwargs...)
+
+**AutoGrad method:** Neither the high level nor the low level interface is sometimes
+appropriate for gradient definitions, e.g. the low level API may not return a value, the
+high level API may have some gradient target parameters as keyword arguments. To solve this
+issue the API exposes an intermediate function with an AutoGrad suffix,
+e.g. `cudnnPoolingForwardAutoGrad`, that is called by the high level method and that makes
```

Comment:
I suggest using the `AD` suffix instead of `AutoGrad`

Reply:
Also needed for the backwards pass